File: AKS.cpp
#include <iostream>
#include <vector>
#include <cln/integer.h>
#include <cln/real.h>
#include <cln/real_io.h>
#include <cln/integer_io.h>
#include <cln/io.h>
#include <cln/input.h>

using namespace std;
using namespace cln;

typedef vector<cl_I> Polynomial;

cl_I order(cl_I r, cl_I n){
	cl_I order = 3;
	while(true){
		if(mod(expt_pos(n, order), r) == 1)
			return order;
		order += 1;
	}
}

cl_I totient(cl_I n){
	cl_I totient;
	for(cl_I i = 1; i < n + 1; i += 1){
		if(gcd(n, i) == 1)
			totient += 1;
	}
	return totient;
}

Polynomial poly_mult_with_mod(Polynomial a, Polynomial b, cl_I r, cl_I n){
	Polynomial product(cl_I_to_ulong(r));
	for(unsigned long long i = 0; i < a.size(); i += 1)
		for(unsigned long long j = 0; j < b.size(); j += 1)
			product[(i + j) % cl_I_to_ulong(r)] = mod(product[(i + j) % cl_I_to_ulong(r)] + (a[i] * b[j]), n);
	return product;
}

Polynomial poly_mod_exp(cl_I a, cl_I r, cl_I n){
	Polynomial xplusa(cl_I_to_ulong(r)), modexp(cl_I_to_ulong(r));
	xplusa[0] = a;
	xplusa[1] = 1;
	modexp[0] = 1;
	unsigned long long power = cl_I_to_ulong(n);
	while(power > 0){
		if(power & 1){
			modexp = poly_mult_with_mod(modexp, xplusa, r, n);
		}
		xplusa = poly_mult_with_mod(xplusa, xplusa, r, n);
		power /= 2;
	}
	return modexp;
}

int main(){
	cl_I n, sqrtn;
	cout<<"Enter an integer > 1 "<<endl;
	cin>>n;
	isqrt(n, &sqrtn);
	// step 1 - checking n == a**b
	for(cl_I a = 2; a < sqrtn + 1; a += 1){
		cl_I num = a;
		for(cl_I b = 2; b < n; b += 1){
			num = num*a;
			if(num == n){
				cout<<"Composite\n";
				return 0;
			}
			else if(num > n)
				break;
		}
	}

	// step 2 - finding smallest r
	cl_R logn = log(n, 2);
	cl_R logsquaredn = (logn)*(logn);
	cl_I r = 3;
	while(true){
		if(gcd(n, r) == 1 && order(r, n) >= logsquaredn)
			break;
		r += 1;
	}

	// step 3 check if 1 < (a, n) < n
	for(cl_I a = 2; a < r + 1; a += 1){
		if(1 < gcd(a, n) && gcd(a, n) < n){
			cout<<"Composite\n";
			return 0;
		}
	}

	// step 4 if n <= r, output prime
	if(n <= r){
		cout<<"Prime\n";
		return 0;
	}

	// step 5 loop iterations to check compositeness
	cl_I termcondtn = floor1(sqrt(totient(r)) * logn);
	for(cl_I a = 1; a <= termcondtn; a += 1){
		cout<<"a: "<<a<<endl;
		Polynomial xplusapown(cl_I_to_ulong(r)), xnplusa(cl_I_to_ulong(r));
		xplusapown = poly_mod_exp(a, r, n);
		xnplusa[cl_I_to_ulong(mod(n, r))] = mod(xnplusa[cl_I_to_ulong(mod(n, r))] + 1, n);
		xnplusa[0] += mod(xnplusa[0] + a, n);

		for(unsigned long i = 0; i < cl_I_to_ulong(r); i++){
			if(i < r-1 && xplusapown[i] != 0)
				cout<<xplusapown[i]<<"x^"<<i<<" ";
			else if(xplusapown[i] != 0)
				cout<<xplusapown[i]<<"x^"<<i;
		}
		cout<<endl;

		for(unsigned long i = 0; i < cl_I_to_ulong(r); i++){
			if(xplusapown[i] != xnplusa[i]){
				cout<<"Composite\n";
				return 0;
			}
		}
	}

	// step 6 output prime
	cout<<"Prime\n";
	return 0;
}

File: AKS.py
from copy import deepcopy
from fractions import gcd
from math import ceil, log, sqrt

def isqrt(n):
	if n < 0:
		raise ValueError('Square root not defined for negative numbers')
	nn = long(n)
	if nn == 0:
		return nn
	q, r = divmod(nn.bit_length(), 2)
	n = 2 ** (q + r)
	while True:
		o = (n + nn // n) // 2
		if o >= n:
			return n
		n = o

def order(r, n):
	o = 3
	while True:
		if pow(n, o, r) == 1:
			return o
		o += 1

def totient(r):
	tot = 0
	for i in xrange(1, r + 1):
		if gcd(r, i) == 1:
			tot += 1
	return tot

def polymul(polya, polyb, n):
	deg = len(polya) + len(polyb)
	poly = [0 for i in xrange(deg)]
	for i in xrange(len(polya)):
		for j in xrange(len(polyb)):
			poly[i + j] += polya[i] * polyb[j]
			poly[i + j] %= n
	return poly_mod(poly, n)

def degree(poly):
	while poly and poly[-1] == 0:
		# from the last
		poly.pop() 
	return len(poly) - 1

def normalize(poly):
	while poly and poly[-1] == 0:
		poly.pop()
	if poly == []:
		poly.append(0)

def polydiv(num, den):
	num = num[:]
	normalize(num)
	den = den[:]
	normalize(den)

	if len(num) >= len(den):
		#Shift den towards right so it's the same degree as num
		shiftlen = len(num) - len(den)
		den = [0] * shiftlen + den
	else:
		return num

	quot = []
	divisor = float(den[-1])
	for i in xrange(shiftlen + 1):
		#Get the next coefficient of the quotient.
		mult = num[-1] / divisor
		quot = [mult] + quot

		#Subtract mult * den from num, but don't bother if mult == 0
		#Note that when i==0, mult!=0; so quot is automatically normalized.
		if mult != 0:
			d = [mult * u for u in den]
			num = [u - v for u, v in zip(num, d)]

		num.pop()
		den.pop(0)

	normalize(num)
	return num

def poly_mod(poly, n):
	for i in range(len(poly)):
		poly[i] %= n
	return poly

def polydiv_with_mod(a, r, n):
	xplusa = [a, 1]
	poly = deepcopy(xplusa)
	xrminusone = [0 for i in xrange(r + 1)] # rth degree
	xrminusone[0] = -1
	xrminusone[r] = 1
	for i in xrange(2, n + 1):
		poly = polymul(poly, xplusa, n)
		if degree(poly) >= r:
			poly = polydiv(poly, xrminusone)
	return poly_mod(poly, n)

def main(n):
	# step 1 - checking n == a**b
	for a in xrange(2, isqrt(n) + 1):
		num = a
		for b in xrange(2, n):
			num *= a
			if num == n:
				return False
			if num > n:
				break

	# step 2 - finding smallest r
	logn = log(n, 2)
	logsquaredn = logn * logn

	r = 3
	while True:
		if gcd(r, n) == 1 and order(r, n) >= logsquaredn:
			break
		r += 1

	# step 3 check if 1 < (a, n) < n
	for a in xrange(2, r + 1):
		if 1 < gcd(a, n) < n:
			return False

	# step 4 if n <= r, output prime
	if n <= r:
		return True

	# step 5 loop iterations to check compositeness
	xnplusa = [0 for i in xrange(n + 1)] # nth degree
	xnplusa[n] = 1 # set the last index
	xrminusone = [0 for i in xrange(r + 1)] # rth degree
	xrminusone[0] = -1
	xrminusone[r] = 1
	for a in xrange(1, long(ceil(sqrt(totient(r)) * logn))):
		polycoeff = deepcopy(xnplusa)
		polycoeff[0] = a
		newpoly = polydiv_with_mod(a, r, n)
		normalize(newpoly)
		polycoeff = poly_mod(polydiv(polycoeff, xrminusone), n)
		normalize(polycoeff)
		print polycoeff, len(polycoeff)
		if cmp(polycoeff, newpoly) != 0:
			return False

	# step 6 output prime
	return True

if __name__ == '__main__':
	n = long(raw_input("Enter an integer > 1 "))
	if main(n):
		print "Prime"
	else:
		print "Composite"